### PR TITLE
Fixed default_columns_for_object when object has non-standard foreign keys

### PR DESCRIPTION
**Pull request by Fivell**
refs #1240

### CODE REVIEW

**Comment by Fivell** [comment]
@mikz , failing test added

**Comment by Fivell** [comment]
I will add next commit to this PR to fix this

**Comment by Fivell** [comment]
@mikz , please review

File: spec/helpers/inputs_helper_spec.rb [hunk 1]
```diff
@@ -398,7 +398,34 @@
       it 'should not render timestamps inputs by default' do
         expect(output_buffer).not_to have_tag('form > fieldset.inputs > ol > li.datetime')
       end
-    
+
+      context "with non-standard foregin keys" do
+        before do
+          #expect(::Post).to receive(:content_columns).and_call_original
```
Comment (mikz) [inline spec/helpers/inputs_helper_spec.rb]: This comment should not be necessary.

Comment (Fivell) [inline spec/helpers/inputs_helper_spec.rb]: @mikz , sure

**Comment by Fivell** [comment]
@mikz , done, anything else ?

File: lib/formtastic/helpers/inputs_helper.rb [hunk 2]
```diff
@@ -350,12 +350,21 @@ def association_columns(*by_associations) # @private
         end
       end
 
+      # Collects non-standard foreign key columns
+      def foreign_key_columns
+        if @object.present? && @object.class.respond_to?(:reflections)
+          @object.class.reflections.map {|_,v| v.options[:foreign_key] }.compact
```
Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: Shouldn't this use just `#foreign_key` method? Because I guess we don't want to show fields for associations with no options.

Also imo it should check just `ActiveRecord::Reflection::BelongsToReflection` as the other foreign keys will not exist on this table. That should be easily checked with `Reflection#belongs_to?`

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz , 

> Shouldn't this use just #foreign_key method? 

for doing this I should change all the doubles to allow foreign_key method, do you want this? 

>> Because I guess we don't want to show fields for associations with no options. 

in this case options will be empty hash and ```compact``` will ignore such columns . Also  we do care only for non-standard foreign keys which are described using ```:foreing_key```  option. 

>>Also imo it should check just ActiveRecord::Reflection::BelongsToReflection as the other foreign keys will not exist on this table. That should be easily checked with Reflection#belongs_to?

is this is ok for you ?

```ruby
 def foreign_key_columns
     if @object.present? && @object.class.respond_to?(:reflections)
         @object.class.reflections.select{|_, v| v.macro == :belongs_to }.map {|_,v|  v.options[:foreign_key] }.compact
     else
        []
     end
   end
```

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: I think it would be less magical to just use `Reflection#foreign_key` which will return the option or it determines the name. That sounds more robust than just checking the options as it will always try to return something even if it is name generated by rails.

Looks like there is a method from Rails `reflect_on_all_associations` that accepts macro. So it can be further simplified:

```
reflections = @object.class.try(: reflect_on_all_associations, :belongs_to) || []
reflections.map { |r| r.foreign_key.to_sym }
```

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz , ok but I need to allow foreign_key for all  the doubles for this

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: Maybe would make sense to use correct instances of `ActiveRecord::Reflection::BelongsToReflection` and not mock them at all. That would allow checking that it works with different rails versions.

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz , 

> 
Maybe would make sense to use correct instances of ActiveRecord::Reflection::BelongsToReflection

Can you add more details, I didn't get your point

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: > @object.class.try(: reflect_on_all_associations, :belongs_to)

As I understand I also should stub reflect_on_all_associations to each test for this. Because specs are working with non-activerecord objects

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: Or include the reflection methods to those models. They are in `ActiveRecord::Reflection::ClassMethods`.

What I meant was create instance like `ActiveRecord::Reflection::BelongsToReflection.new` instead of creating mocks. That way we can really guarantee rails compatibility.

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz , there is no activerecord dependency in tests now, not sure it is good idea to add it for such a small improvement

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: It is quite a big improvement because it ensures this will work with different rails versions. What if rails changes format of that object or that method? Because it is a mock no one can know because it does not test the rails interface. 

The tests depend on rails. They include methods from ActionView, ActiveModel and other parts. Using object from ActiveRecord and include few methods is not going to include more dependencies.

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: I see what you mean. Formtastic actually does not depend on ActiveRecord. But then it is kind of weak dependency as those methods like `content_columns` are from ActiveRecord.

I'd say it is fine to add ActiveRecord as a test dependency, because we actually want to test against ActiveRecord. I'll start a branch soon exploring that and replacing few mocks with ActiveRecord objects.

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz, 

>> What if rails changes format of that object or that method? Because it is a mock no one can know because it does not test the rails interface.


generally I agree with you about test changes,  however  formtastic tests already over-mocked and over-stubbed and needs more efforts and time to add test for each edge case than it deserves.

>>  I'll start a branch soon exploring that and replacing few mocks with ActiveRecord objects.

feel free to ping me when it will be more clear how to arrange tests after this, so I could finish it if I have free time.

Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: Yes formtastic's test suite is totally over-mocked and over-stubbed. I'll work over time to clean it up. 

https://github.com/justinfrench/formtastic/pull/1242 is first spark of integrating ActiveRecord.

I think you could just create some new models and add them to the schema to use real ActiveRecord objects.

**Comment by Fivell** [comment]
blocked by #1241 until will be merged in

**Comment by mikz** [comment]
@Fivell merged it.

**Comment by Fivell** [comment]
@mikz ,  changed and  rebased

File: lib/formtastic/helpers/inputs_helper.rb [hunk 3]
```diff
@@ -350,12 +350,21 @@ def association_columns(*by_associations) # @private
         end
       end
 
+      # Collects all foreign key columns
+      def foreign_key_columns
```
Comment (mikz) [inline lib/formtastic/helpers/inputs_helper.rb]: I guess this should be marked as ` # @private` to not show in the yard documentation

Comment (Fivell) [inline lib/formtastic/helpers/inputs_helper.rb]: @mikz , done

**Comment by mikz** [comment]
@Fivell thank you very much! This is good change and good improvement in testing against real ActiveRecord!

Will merge it right after build is green.

**Comment by Fivell** [comment]
@mikz,  thanks for help in fixing it 🚀
